common.go

```diff
@@ -7,9 +7,6 @@ type Pair[K any, V any] interface {
 
 	// Value Returns the value associated with the pair. Will return the zero value for V if IsEmpty is true.
 	Value() V
-
-	// IsEmpty returns true iif the pair is empty.
```

Comment:
I really like just being able to return 0 and have things work out.

What is the benefit of removing this? How does it help expressiveness, or understanding ?

Reply:
Using the pattern to return `(V, bool)` is consistent with map behavior so it's a similar pattern that devs will know.
The proposed change guarantees that a `Pair` is always a real K-V. I do think that simplification makes code simpler to reason about since it means code never needs to "worry" about `IsEmpty` being true. This is useful when passing a Pair to other functions.

Reply:
The change to 'Find' seems ok to me , but  I'm not a big fan of the change to `LeastUpperBound` and `GreatestLowerBound`.

I think they are simpler by having the IsEmpty be a property of the pair, rather than returning `(Pair[K,V], bool)`.

Maybe I've spent too much time in C++.... 

But I think the `GreatesLowerBound`, `LeastUpperBound`, `Least` and `Most` methods are easier to work with the original interface then with the proposed changes.

And I'd prefer the consistency of having `Find` use a similar interface.

I'm inclined to say "no" to this change, but I want to give you the chance to tell me why I'm wrong.

Reply:
Ok. A challenge; i'll give it a shot...

I do agree that the `Greatest/Least/Find` set of functions should be similar. `Find` is a bit different since you know exactly the key you are finding; not so for `Greatest/Least`. But I agree that the "shape" of the return value should be similar.

I think the right method to optimize for is `Find`.
Picking random go [code](https://github.com/golang/go/blob/master/src/cmd/compile/internal/abt/avlint32.go) that uses/implements a tree; `Find` is used way more than `Glb` and `Lub`. I should probably check more than just one repo but it "feels" likely for most applications need `Find` more than `Lub` and `Glb`.

Consider:
```go
  m := map[string]int{}
  ...
  v, ok := m["item"]
  if !ok { .... }
```

It is a common pattern in go. The analogous code is:
```go
   t := EmptyTree[string, int]{}
   ...
   v, ok := t.Find("item")
   if !ok {... }
```

It is perhaps a personal style but I like the symmetry between the two collections here. I think an individual reading the tree code who is familiar with go will just "know" what it means.

In fact, on a slightly unrelated note, I was thinking about this more; and I wonder why does `Pair` need to be an interface; as opposed to a struct.  Making it a struct means it is never nil which is reasonable  IMO b/c if `Find`/`Glb`/`Lub` do find a KV to return then you are guaranteed it is not nil. This is similar to returning a `(structType, error)` from a function. `Pair` is really a container for a KV so I don't foresee multiple implementations for it.

Reply:
I think you do make a super reasonable argument about trying to be idiomatic.

One key difference, however, is that maps in go by default optimize for the case where the zero value in the map denotes absence. That is, the simplest way to use a map is:

```
v := m["item"]
if v != nil {
   //
}
```

For maps, go supports the `ok` case as an exception: use this extended syntax in the event the default assumptions don't apply.

It doesn't force you to do

```
v, _ := ...
```

In the common case, which is much more predominant.

In the cases where errors are returned you almost always need to do something with the error (typically return it and bail out). In fact for error returning functions the syntax `v, _ := someFunctionThatReturnsAndError()` is usually a sign that the code is doing something unusual and it draws extra attention to it.

I think forcing to use ok all the time when it usually isn't needed doesn't make sense.

What about the following as a compromise:

1. `Find(key K) V`
2. `FindOpt(key K) (V, bool)`
3. `LeasetUpperBound(key K) Pair[K,V]`
4. `LeastUpperBondOpt(key K) (Pair[K,V], bool)`
5. `...`

We can then go ahead and make Pair a struct not an interface.

This way the common case is simple, and we don't have to litter code with a bunch of extra `_`.

Reply:
Ok. Maybe a further refinement..
Instead of `Find` as you define above, what about:
1. `Find` -> `MustFind` and will panic when not found since no caller should call this w/o expecting to find something.
2. `FindOpt` -> `Find`
3. `LeastUpperBound` -> `MustLeastUpperBound`
4. `LeastUpperBoundOpt` -> `LeastUpperBound`

I do think this achieves the goal of having an operation to search the tree without the added "littering" of `_` and even makes it explicit with `Must` which is a pattern also used by `regexp.MustCompile` or `template.Must`

Reply:
FYI: i fixed the broken test

Reply:
Go only panics in very unexpected situations. For example, `new` or `make` failing.

An object not existing in a map is not something that should cause a  panic.

I am a firm No on `MustFind`.

The default experience when using a “map like interface” should be “the zero value implies ‘does not exist’.

Any complexity around dealing with the difference between “zero value” and “does not exist” should not impact the mainline case.

The default design mantra should be “simple things are simple; complex things are possible”.

Reply:
Ok; panic for not found does seem extreme. I'm not a fan of `FindOpt` -- I feel no one will know what it means. Let's add a `Contains` or `Has` method that returns bool and call it good?

Reply:
I added `Contains(key K) bool` method and defined `Find(key K) V`
I left `Lub`, `Glb` as is since we agreed to optimize for the more commonly used `Find`
Let me know your thoughts -- next PR makes `Pair` struct